File: go/client/authenticator.go
package client

import (
	"fmt"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/weft-finance/vayu-go/openapi"
)

const tokenExpiryThreshold = 5 * time.Minute

type Authenticator struct {
	apiKey      string
	accessToken string
	expiresAt   time.Time
	apiClient   *openapi.APIClient
}

func NewAuthenticator(apiKey string, apiClient *openapi.APIClient) *Authenticator {
	return &Authenticator{
		apiKey:    apiKey,
		apiClient: apiClient,
	}
}

func (a *Authenticator) IsAuthenticated() bool {
	return a.accessToken != ""
}

func (a *Authenticator) AccessToken() string {
	return a.accessToken
}

func (a *Authenticator) NeedsRefresh() bool {
	return !a.IsAuthenticated() || time.Now().Add(tokenExpiryThreshold).After(a.expiresAt)
}

// EnsureValidToken returns a valid access token, logging in or refreshing as needed.
func (a *Authenticator) EnsureValidToken() (string, error) {
	if a.NeedsRefresh() {
		if err := a.Authenticate(); err != nil {
			return "", err
		}
	}

	return a.accessToken, nil
}

// Authenticate exchanges the refresh token for a new access token.
func (a *Authenticator) Authenticate() error {
	ctx, cancel := GenerateContextWithTimeout()
	defer cancel()

	request := a.apiClient.AuthAPI.Login(ctx)
	request = request.LoginRequest(openapi.LoginRequest{RefreshToken: a.apiKey})
	response, _, err := request.Execute()

	if err != nil {
		return BuildVayuErrorFromGenericOpenAPIError(err)
	}

	a.accessToken = response.AccessToken

	if err := a.extractJWTExpiry(); err != nil {
		return err
	}

	return nil
}

func (a *Authenticator) extractJWTExpiry() error {
	token, _, err := new(jwt.Parser).ParseUnverified(a.accessToken, jwt.MapClaims{})
	if err != nil {
		return BuildVayuError(fmt.Errorf("failed to parse JWT"))
	}

	claims, ok := token.Claims.(jwt.MapClaims)
	if !ok {
		return BuildVayuError(fmt.Errorf("failed to parse JWT claims"))
	}

	exp, ok := claims["exp"].(float64)
	if !ok {
		return BuildVayuError(fmt.Errorf("expiry date not found in JWT"))
	}

	a.expiresAt = time.Unix(int64(exp), 0)
	return nil
}

File: python/vayu_sdk/clients/authenticator.py
from openapi.api_client import ApiClient
from openapi.configuration import Configuration
from openapi.api.auth_api import AuthApi
from openapi.models.login_request import LoginRequest
import jwt
from datetime import datetime, timedelta


class VayuAuthenticator:
    _TOKEN_EXPIRY_THRESHOLD = timedelta(minutes=5)

    def __init__(self, api_key: str, host: str = None):
        self._api_key = api_key
        self._host = host
        self._access_token: str = None
        self._expires_at: datetime = datetime.min

    @property
    def is_authenticated(self) -> bool:
        return self._access_token is not None

    @property
    def access_token(self) -> str:
        return self._access_token

    @property
    def needs_refresh(self) -> bool:
        return (
            self._access_token is None
            or datetime.now() + self._TOKEN_EXPIRY_THRESHOLD > self._expires_at
        )

    def ensure_valid_token(self) -> str:
        """Return a valid access token, logging in or refreshing as needed."""
        if self.needs_refresh:
            self.authenticate()

        return self._access_token

    def authenticate(self):
        """Exchange the refresh token for a new access token."""
        configuration = Configuration(host=self._host)
        client = ApiClient(configuration)
        api = AuthApi(client)
        login_request = LoginRequest(refreshToken=self._api_key)
        response = api.login(login_request)

        self._access_token = response.access_token

        decoded_token = jwt.decode(
            self._access_token, options={"verify_signature": False}
        )
        self._expires_at = datetime.fromtimestamp(decoded_token["exp"])


__all__ = ["VayuAuthenticator"]
